source/common/stats/thread_local_store.h

```diff
@@ -16,6 +16,98 @@
 namespace Envoy {
 namespace Stats {
 
+/**
+ * A histogram that is stored in TLS and used to record values per thread. This holds two
+ * histograms, one to collect the values and other as backup that is used for merge process. The
+ * swap happens during the merge process.
+ */
+class ThreadLocalHistogramImpl : public Histogram, public MetricImpl {
+public:
+  ThreadLocalHistogramImpl(const std::string& name, std::string&& tag_extracted_name,
+                           std::vector<Tag>&& tags);
+
+  ~ThreadLocalHistogramImpl();
+
+  void merge(histogram_t* target);
+
+  /**
+   * Called in the beginning of merge process. Swaps the histogram used for collection so that we do
+   * not have to lock the histogram in high throughput TLS writes.
+   */
+  void beginMerge() {
+    // This switches the current_active_ between 1 and 0.
+    current_active_ = otherHistogramIndex();
+  }
+
+  // Stats::Histogram
+  void recordValue(uint64_t value) override;
+  bool used() const override { return flags_ & Flags::Used; }
+
+private:
+  uint64_t otherHistogramIndex() const { return 1 - current_active_; }
+  uint64_t current_active_;
+  histogram_t* histograms_[2];
+  std::atomic<uint16_t> flags_;
+  std::thread::id created_thread_id_;
+};
+
+typedef std::shared_ptr<ThreadLocalHistogramImpl> TlsHistogramSharedPtr;
+
+/**
+ * Class used to create ThreadLocalHistogram in the scope.
+ */
+class TlsScope : public Scope {
```

Comment:
Is there any reason we can't just add `tlsHistogram()` to the `Scope` interface and remove this class?

Reply:
As per the general guideline we are taking in this design about not exposing ThreadLocal's  to stats interface, I have added this class to specifically to have that method.

Reply:
@mattklein123 I will defer this to a follow-up PR. added TODO.

Reply:
@mattklein123 can you point me to code samples that use direct TLS if the user is in thread local context? I will look at them and see If I can make that change?

Reply:
@ramaraochavali any pattern like here: https://github.com/envoyproxy/envoy/blob/master/source/common/http/codes.cc#L82 where we dynamically grab a histogram and immediately record it. We could skip a double TLS lookup in this case if we offered a tlsHistogram() accessor. I think it would be pretty easy to do, but let's obviously sort out all of our other issues first. :) (Maybe open an issue to track.)

Reply:
Sure :-). With the finding we had yesterday I am hoping that they would comeback soon with the solution.